### ISSUE TITLE
Make graph visible in web UI

### ISSUE THREAD

**Issue by easbar** [graphhopper/graphhopper#1456]
I think for debugging and analysis of our routing algorithms it would be incredibly useful to make the graph as visible as possible. I think for some use-cases drawing the graph directly on top of the OSM layer also has several advantages over the MiniGraph/java-based/Swing approach. I have started to build a little prototype along this line and at the moment it looks like this:

![image](https://user-images.githubusercontent.com/17603532/45562771-a0861580-b84b-11e8-9fdb-4b303f4c91aa.png)

I can draw a bounding box on the map layer and then query the server for the edges of the graph and draw them on the map. I have a few more ideas that would be useful like drawing CH shortcuts, filtering edges by node level (to see the CH contraction evolve), visualize fwd/bwd searches etc.

So for development and debugging I think this will be very useful, but at the same time this should probably not be available on graphhopper.com/maps, so there are a few options:

1.  keep this code in a separate branch (I would try to put as much code as possible in additional files so that its easy to merge master at any time)

2.  add this to master and use some isProduction() flag to disable the additional code (same goes for the REST endpoint I am using on the server)

3.  do this in a separate project (maybe without routing) (probably not a very good idea, because we need to change code on the server as well)

Does anybody else think this would be useful ? Should this be part of this repository in any way ? I guess as long as I am the only one using this I will go with the first option. It will probably take some time to figure out what is the best way to make use of this feature and in many cases there will probably be some quick experimental code to debug a specific feature.

**Comment by karussell** [comment]
Yes, something like this should be done. 

> so there are a few options:

This could be done in the tools module. As tools is not included in the web module. But I do not care much, could be also outside.

What is the performance? With leaflet I made bad experience. I made good experience with [deck.gl](https://github.com/graphhopper/isochrone-experiments) even for huge areas.

**Comment by ammagamma** [comment]
> What is the performance?

Performance is definitely an issue at the moment especially when I add labels for edge and node ids (which I think is essential). For very small bounding boxes it is still fine, but already for a box as big as the one in the screenshot the rendering becomes very slow when I add the labels. To debug routing bugs we can in almost any case reduce the map to a small part, so the current version could still be useful, but I would prefer not to be limited like this.

Have you tried any approaches with leaflet where only lines/labels are rendered that are actually visible for the current zoomlevel . It would be fine to hide (or cluster) the labels when we zoom out because it is impossible to distinguish single nodes and edges without zooming in deeply anyway. I briefly looked into this but could not find a working solution that improves performance significantly. I think it would be useful to re-use the existing leaflet web UI, because I can use the routing feature, see instructions etc.

I also looked at your deck.gl experiments and it is much faster and the performance was still very good when I added a text label to every edge. Right now it looks to me as if this might be the easier solution to fix performance issues. Do we have to use mapTiler or could we get the OSM tiles from omniscale, openstreetmap.org etc. like in graphhopper/maps ?

**Comment by karussell** [comment]
> Have you tried any approaches with leaflet where only lines/labels are rendered that are actually visible for the current zoomlevel ? 

Yes, not recently but I tried this too at some point. At that time I also tried to paint on a canvas directly but this was not that easy and still not much faster (plus various bugs ;)).

> I think it would be useful to re-use the existing leaflet web UI

We can try to switch only the map part ;D
For deck.gl this might be not possible, but maybe mapbox GL JS works and for me it was a bit faster than leaflet.

For deck.gl we could implement even "mouse over" or click **single** edges/nodes and then show information, labels or neighboring nodes. Something granular makes leaflet even slower.

> Do we have to use mapTiler or could we get the OSM tiles from omniscale, 

Vector tiles from various providers work but osm or omniscale do not offer them I think, still the style of the vector tiles can be highly customized. (For osm.org vector tiles are planned)

**Comment by karussell** [comment]
Another idea in the experiments was to serve the graph as vector tiles but I couldn't get it working and the indicated speed was much slower than the compact binary format.

**Comment by ammagamma** [comment]
> Yes, not recently but I tried this too at some point. At that time I also tried to paint on a canvas directly but this was not that easy and still not much faster (plus various bugs ;)).

ok I take this off my list then ;)

> We can try to switch only the map part ;D

You mean removing leaflet from current web UI  and replace it with one of the other solutions ? Do you think that isochrones will not work properly with leaflet (#1453) ? Why do you think it is not possible with deck.gl ? 

> For deck.gl we could implement even "mouse over" or click single edges/nodes

For the debugging usecases I think this would be very important. This also works with leaflet.

**Comment by ammagamma** [comment]
> Another idea in the isochrone-experiments repo was to serve the graph as vector tiles but I couldn't get it working and the indicated speed was much slower than the compact binary format.

For debugging it would be best to be as flexible as possible (quickly add some property to the server response and use it in javascript / change the visualization on the map), so I guess drawing/adjusting the graph in the browser would be better from this point of view.

**Comment by karussell** [comment]
> ok I take this off my list then ;)

You can still try. It was a long time ago and also have no real JS skills. Or I can ask for help elsewhere.

> You mean removing leaflet from current web UI and replace it with one of the other solutions ? 

yes

> Do you think that isochrones will not work properly with leaflet (#1453) ? 

No, this issue is just about to create a dedicated UI for isochrones. I do not want to increase the functionality of the existing, already too complex GH Maps

> Why do you think it is not possible with deck.gl ?

deck.gl is based on react JS, completely different dev environment than JS but yes, probably possible with more experience :)

> so I guess drawing/adjusting the graph in the browser would be better from this point of view.

What do you mean here? Vector tiles are very flexible. You can put lots of information there and only pick the things you like

**Comment by ammagamma** [comment]
> deck.gl is based on react JS, completely different dev environment than JS but yes, probably possible with more experience :)

I think deck.gl can also work without react: https://github.com/uber/deck.gl/blob/master/docs/get-started/using-standalone.md, but using react seems like the way they advertise.

> What do you mean here? Vector tiles are very flexible. You can put lots of information there and only pick the things you like

I thought about adjusting the visualization of the graph in the browser like a slider that filters nodes and edges by some properties (like the CH level) and this sounded like the tiles would be fixed on the server side, but maybe I have a false conception of this approach.

So if isochrones should not go int GH Maps I guess there is no real reason to remove leaflet from it and I think I will just go with a little react/deck.gl app like your isochrone prototype to visualize the graph. I might either add some minimal version for routing or just open graphhopper/maps in a second browser window to see the route.

**Comment by michaz** [comment]
> I can draw a bounding box on the map layer and then query the server for the edges of the graph and draw them on the map.

@ammagamma 
Wait, really? You put a "real" bounding-box query into `LocationIndexTree`, or how are you doing this? I've been waiting for that..

**Comment by boldtrn** [comment]
Instead of switching to deck.gl we could also have a look at [OpenLayers](https://openlayers.org/). OpenLayers supports vector tiles and raster tiles, so I think we could keep the current functions of GH maps, but simply add a vector overlay that contains the graph data.

@ammagamma do you know where the bottleneck is? Is the rendering slow, or the generation on the server-side?

If it's just the rendering, I think serving GeoJson/TopoJson/mvt and use a vector tile client could make the the rendering very fast.

**Comment by michaz** [comment]
@ammagamma 
Meaning, I'm really excited about that, and it would close at least one of each mine and @karussell's tickets. :)

**Comment by ammagamma** [comment]
> You put a "real" bounding-box query into LocationIndexTree, or how are you doing this? I've been waiting for that..

No, sorry, I was doing this the trivial way just iterating over all edges and filtering out those that lie in the box, but obviously an efficient query taking a bbox as input would be useful here as well.

**Comment by boldtrn** [comment]
> yes, it is definitely the rendering that is slow and with deck.gl it is running much faster.

Ok, I think this should be solvable then :).

Have a look at [Maputnik](https://maputnik.github.io/editor/#6.61/48.725/10.086), there is an inspect mode, that renders raw vector data. I think something like this would be awesome for GraphHopper and to debug the Graph. We could think about how we could offer a more or less similar endpoint to a GraphHopper instance and render "graph tiles". As you can see, the rendering is very fast and the amount of data is comparable or more than the amount of data we need for representing graphs.

![maputnik](https://user-images.githubusercontent.com/1553525/45631271-d967f980-ba9a-11e8-8e90-0319701cb204.png)

BTW: [Maputnik is open source](https://github.com/maputnik/editor)

**Comment by karussell** [comment]
>  @karussell did you compare OpenLayers with deck.gl when you did the isochrone experiments ?

No

**Comment by ammagamma** [comment]
If you want to have a look, I put a little tool [here](https://github.com/ammagamma/graphhopper-graphtool) for now. It is using deck.gl (without React) at the moment. It is still very experimental, but it already has a few useful (for me at least) features:

- draws all nodes & edges that lie in a bounding box that you can select on a map
- shows node and edge ids on the map (you need to zoom in far enough)
- tooltip for nodes and edges to display some details (like edge weight)
- bonus (as proof of concept): there is a slider that filters by CH level, if you move it from left to right you can see the order in which nodes got contracted (I think/hope things like this can be very useful to understand some aspects of CH and other algorithms better).

The performance is definitely sufficient for medium sized cities (like Nuremberg) and I haven't really paid attention to any kind of optimizations. For larger cities the rendering starts to become annoyingly slow, but this depends on what you are trying to do I guess. 

The tool queries the server for the graph data (as json) and then renders vector data on top of the graph, I did not look into any approches using vector tiles so far. I tried changing the map back to leaflet and use deck.gl only to draw the vector layer, but I could not make it so that the vector layer and the map position stay in sync. I also tried OpenLayers (its in the open-layers folder in the same repository), but I quickly hit a performance problem when I added text labels. OpenLayers [docs](http://openlayers.org/en/latest/doc/tutorials/background.html) say:
> By default, OpenLayers uses a performance optimized Canvas renderer. An experimental WebGL renderer (without text rendering support) is also avaialble.

Maybe this can be fixed, but I stopped trying after this.

**Comment by boldtrn** [comment]
Thanks for the details, I will try to have a look :).

> I also tried OpenLayers (its in the open-layers folder in the same repository), but I quickly hit a performance problem when I added text labels. OpenLayers docs say

I am sorry if my comments about OpenLayers cost you too much time :(.

**Comment by ammagamma** [comment]
> I am sorry if my comments about OpenLayers cost you too much time :(

No worries, I think it is certainly worth a look and it is very hard to know what suites our needs better without trying (I still do not know).

**Comment by karussell** [comment]
:+1: 

> For larger cities the rendering starts to become annoyingly slow ...
> The tool queries the server for the graph data (as json) ...

As collecting the edges fro the graph should be still fast, maybe this is because the transmitted data becomes huge (>30MB)? Did you try to implement a simple binary format and gzip it? See [this prototype](https://github.com/graphhopper/graphhopper/blob/ceaba060d8a9e050d6dcf568e9bbbae787da4b88/web-bundle/src/main/java/com/graphhopper/resources/IsochroneResource.java#L117).

**Comment by easbar** [comment]
>  Did you try to implement a simple binary format and gzip it? 

No, but the rendering performance is not an issue for what I wanted to do (not since I was using deck.gl).

**Comment by karussell** [comment]
I mean this more for the bandwidth. For me the problem was without compression that it was like 5 times larger and so after using the binary format it was more responsive.

**Comment by easbar** [comment]
Ah sorry yes I meant data-transfer+rendering time not just rendering which is rather unrelated to the compression I guess (?). Yes I am sure for larger graphs or isochrones etc. the compression is important. Actually I did start implementing this, but did not try it so far I could tell whether or not its an improvement (it takes a bit more effort and since I was happy with the response/rendering time I did not see much value). See https://github.com/easbar/graphhopper/tree/graphtool-binary-format and https://github.com/easbar/graphhopper-graphtool/tree/binary_format.
